### PR TITLE
[8.1] feat: RequestManagementSystem: remove DN in favour of owner

### PR DESCRIPTION
**Pull request by fstagni**
- [x] add to wiki

BEGINRELEASENOTES

*RMS
CHANGE: remove DN in favor of owner

ENDRELEASENOTES

### CODE REVIEW

**Comment by chaen** [comment]
Updating the DB will not be enough. You will also need a way to take care of the requests in the proxy DB, or those coming from already running jobs. They will still have `ownerDN` as attribute, and `SQLAlchemy` will likely not like that on the server side.

**Comment by fstagni** [comment]
Good point. Then we should keep both Owner and OwnerDN in DB, meaning adding "Owner" here without removing OwnerDN. Agree?

**Comment by chaen** [comment]
Absolutely. And *consume* only the owner

**Comment by fstagni** [comment]
Added back the `OwnerDN`. Now, unless I code for backward compatibility (meaning adding `if OwnerDN and not Owner` etc.. ) this means that a possible script for updating the DB should be run more than once after the release is installed. Acceptable?

**Comment by chaen** [comment]
I'd rather go with the `if OwnerDN and not Owner` in `getRequest`. It's an easy enough change. And we can immediately do the PR for removing the hack for the next version.

**Comment by fstagni** [comment]
> I'd rather go with the `if OwnerDN and not Owner` in `getRequest`. It's an easy enough change. And we can immediately do the PR for removing the hack for the next version.

Done, but TBH not fully sure about it.

File: src/DIRAC/DataManagementSystem/Client/FTS3Operation.py [hunk 1]
```diff
@@ -372,7 +372,7 @@ def fromRMSObjects(cls, rmsReq, rmsOp, username):
 
         :param rmsReq: RMS Request object
         :param rmsOp: RMS Operation object
-        :param username: username to which associate the FTS3Operation (normally comes from the Req OwnerDN)
+        :param username: username to which associate the FTS3Operation (normally comes from the Req Owner)
```
Comment (andresailer) [inline src/DIRAC/DataManagementSystem/Client/FTS3Operation.py]: If this comes from the request now, why not make the username argument optional, or drop it, and use `rmsReq.Owner` inside the function?

File: src/DIRAC/TransformationSystem/Client/RequestTasks.py [hunk 2]
```diff
@@ -118,7 +110,7 @@ def _multiOperationsBody(self, transJson, taskDict, ownerDN, ownerGroup):
             For example ``TASK:TargetSE`` is replaced with ``task['TargetSE']``
 
         :param dict taskDict: dictionary of tasks, modified in this function
-        :param str ownerDN: certificate DN used for the requests
+        :param str owner: certificate DN used for the requests
```
Comment (andresailer) [inline src/DIRAC/TransformationSystem/Client/RequestTasks.py]:
```suggestion
        :param str owner: owner used for the requests
```

File: src/DIRAC/TransformationSystem/Client/test/Test_Client_WorkflowTasks.py [hunk 3]
```diff
@@ -91,9 +91,11 @@
         (taskDictNoInputsNoSite, True, True, expectedBulk),
     ],
 )
-def test_prepareTranformationTasks(taskDictionary, bulkSubmissionFlag, result, expectedRes):
+def test_prepareTranformationTasks(mocker, taskDictionary, bulkSubmissionFlag, result, expectedRes):
+    mocker.patch("DIRAC.TransformationSystem.Client.WorkflowTasks.getDNForUsername", return_value=S_OK(["pippo"]))
```
Comment (andresailer) [inline src/DIRAC/TransformationSystem/Client/test/Test_Client_WorkflowTasks.py]:
```suggestion
    mocker.patch("DIRAC.TransformationSystem.Client.WorkflowTasks.getDNForUsername", return_value=S_OK(["/bih/boh/DN"]))
```
Or something that cannot be confused with the username

File: src/DIRAC/WorkloadManagementSystem/Agent/test/Test_Agent_JobCleaningAgent.py [hunk 4]
```diff
@@ -128,6 +128,7 @@ def test_deleteJobOversizedSandbox(mocker, inputs, params, expected):
     mocker.patch("DIRAC.WorkloadManagementSystem.Agent.JobCleaningAgent.JobDB", return_value=mockNone)
     mocker.patch("DIRAC.WorkloadManagementSystem.Agent.JobCleaningAgent.ReqClient", return_value=mockNone)
     mocker.patch("DIRAC.WorkloadManagementSystem.Agent.JobCleaningAgent.JobMonitoringClient", return_value=mockJMC)
+    mocker.patch("DIRAC.WorkloadManagementSystem.Agent.JobCleaningAgent.getDNForUsername", return_value=S_OK(["pippo"]))
```
Comment (andresailer) [inline src/DIRAC/WorkloadManagementSystem/Agent/test/Test_Agent_JobCleaningAgent.py]:
```suggestion
    mocker.patch("DIRAC.WorkloadManagementSystem.Agent.JobCleaningAgent.getDNForUsername", return_value=S_OK(["/bih/boh/DN"]))
```

**Review by andresailer — COMMENTED** [review]
Wow, must have been a pain to change all this!

File: src/DIRAC/DataManagementSystem/Client/FTS3Operation.py (hunk 1 above)
Comment (fstagni) [reply src/DIRAC/DataManagementSystem/Client/FTS3Operation.py]: Good suggestion, I apply it.

File: src/DIRAC/RequestManagementSystem/DB/RequestDB.py [hunk 5]
```diff
@@ -237,8 +236,13 @@ def cancelRequest(self, requestID):
         try:
             updateRet = session.execute(
                 update(Request)
-                .where(Request.RequestID == requestID)
-                .values({Request._Status: "Canceled", Request._LastUpdate: datetime.datetime.utcnow()})
+                .where(Request.RequestID == requestID)  # pylint: disable=no-member
+                .values(
+                    {
+                        Request._Status: "Canceled",  # pylint: disable=no-member
```
Comment (chaen) [inline src/DIRAC/RequestManagementSystem/DB/RequestDB.py:244]: that's a lot of cosmetic changes, which makes it hard to review. 
Why did you add all this `pylint` hints ?

File: src/DIRAC/RequestManagementSystem/DB/RequestDB.py [hunk 6]
```diff
@@ -600,7 +648,7 @@ def getRequestSummaryWeb(self, selectDict, sortList, startItem, maxItems):
             "RequestID",
             "RequestName",
             "JobID",
-            "OwnerDN",
+            "Owner",
```
Comment (chaen) [inline src/DIRAC/RequestManagementSystem/DB/RequestDB.py:652]: I guess you need a corresponding PR for the webapp then ?

File: src/DIRAC/RequestManagementSystem/DB/test/RMSTestScenari.py [hunk 7]
```diff
@@ -22,7 +20,7 @@ def test_stress(reqDB):
 
     reqIDs = []
     for i in range(STRESS_REQUESTS):
-        request = Request({"RequestName": "test-%d" % i})
+        request = Request({"RequestName": "test-%d" % i, "Owner": "pippo"})
```
Comment (chaen) [inline src/DIRAC/RequestManagementSystem/DB/test/RMSTestScenari.py]: why do you add it ? It should be self discovered as before no ?

File: src/DIRAC/RequestManagementSystem/private/RequestValidator.py [hunk 8]
```diff
@@ -276,20 +271,20 @@ def setAndCheckRequestOwner(request, remoteCredentials):
         :returns: True if everything is fine, False otherwise
         """
 
-        credDN = remoteCredentials["DN"]
+        cred = remoteCredentials["username"]
```
Comment (chaen) [inline src/DIRAC/RequestManagementSystem/private/RequestValidator.py]:
```suggestion
        credUsername = remoteCredentials["username"]
```

File: src/DIRAC/RequestManagementSystem/private/RequestValidator.py [hunk 9]
```diff
@@ -1,8 +1,3 @@
-########################################################################
-# File: RequestValidator.py
-# Author: Krzysztof.Ciba@NOSPAMgmail.com
-# Date: 2012/09/18 07:55:16
-########################################################################
 """ :mod: RequestValidator
```
Comment (chaen) [inline src/DIRAC/RequestManagementSystem/private/RequestValidator.py:1]: The `RequestValidator` needs to work with both versions, because it is instantiated on the client and the server side, so the tests should cover both cases

File: src/DIRAC/TransformationSystem/Client/WorkflowTasks.py [hunk 10]
```diff
@@ -106,11 +103,10 @@ def prepareTransformationTasks(
             owner = proxyInfo["username"]
             ownerGroup = proxyInfo["group"]
 
-        if not ownerDN:
```
Comment (chaen) [inline src/DIRAC/TransformationSystem/Client/WorkflowTasks.py:109]: I am not entirely sure the behavior is the same. As it was before, you could run with a given proxy and still set the `ownerDN` to a different identity and I wonder if we do not rely on that somewhere

**Review by chaen — COMMENTED** [review]
I tried to review but it is difficult. It's maybe a good idea to apply this patch on the server side for certification tomorrow

File: src/DIRAC/RequestManagementSystem/DB/RequestDB.py (hunk 5 above)
Comment (fstagni) [reply src/DIRAC/RequestManagementSystem/DB/RequestDB.py:244]: Because there was a `pylint: disable=no-member` applied for the entire file (which I removed, as it is bad practice).

File: src/DIRAC/RequestManagementSystem/DB/test/RMSTestScenari.py (hunk 7 above)
Comment (fstagni) [reply src/DIRAC/RequestManagementSystem/DB/test/RMSTestScenari.py]: Mmm it wasn't... You mean it might be an error?

File: src/DIRAC/TransformationSystem/Client/WorkflowTasks.py (hunk 10 above)
Comment (fstagni) [reply src/DIRAC/TransformationSystem/Client/WorkflowTasks.py:109]: What you mention is, IIRC, a "trick" used for running the TransformationSystem in a multi-VO setup. I defer the answer to @andresailer

**Comment by fstagni** [comment]
> I tried to review but it is difficult. It's maybe a good idea to apply this patch on the server side for certification tomorrow

You mean, merge as-is, try it out, and go back to the comments if needed?

File: src/DIRAC/RequestManagementSystem/DB/RequestDB.py (hunk 6 above)
Comment (fstagni) [reply src/DIRAC/RequestManagementSystem/DB/RequestDB.py:652]: https://github.com/DIRACGrid/WebAppDIRAC/pull/714

File: src/DIRAC/RequestManagementSystem/DB/test/RMSTestScenari.py (hunk 7 above)
Comment (chaen) [reply src/DIRAC/RequestManagementSystem/DB/test/RMSTestScenari.py]: yes, it should work

**Comment by chaen** [comment]
> You mean, merge as-is, try it out, and go back to the comments if needed?

no no, not merge, just `git apply` it

File: src/DIRAC/TransformationSystem/Client/WorkflowTasks.py (hunk 10 above)
Comment (andresailer) [reply src/DIRAC/TransformationSystem/Client/WorkflowTasks.py:109]: I hope this is only used so that tasks are submitted by the owner of the transformation rather than the shifter

File: src/DIRAC/RequestManagementSystem/private/RequestValidator.py (hunk 9 above)
Comment (fstagni) [reply src/DIRAC/RequestManagementSystem/private/RequestValidator.py:1]: Check what I have done for covering this case.

File: src/DIRAC/RequestManagementSystem/private/RequestValidator.py [hunk 11]
```diff
@@ -276,20 +271,31 @@ def setAndCheckRequestOwner(request, remoteCredentials):
...
+            ownershipCheck = request.Owner
+        # ##
+
         # If the owner or the group was not set, we use the one of the credentials
-        if not request.OwnerDN or not request.OwnerGroup:
-            request.OwnerDN = credDN
+        if not ownershipCheck or not request.OwnerGroup:
+            # FIXME: code for backward compatibility with requests created by 8.0 clients
+            if hasattr(request, "OwnerDN"):
+                request.OwnerDN = remoteCredentials["DN"]
```
Comment (chaen) [inline src/DIRAC/RequestManagementSystem/private/RequestValidator.py]: I am not sure I understand why do you need to set the `OwnerDN` if it has that attribute ?

File: src/DIRAC/RequestManagementSystem/private/test/Test_RequestValidator.py [hunk 12]
```diff
@@ -75,12 +75,12 @@ def testValidator(self):
         self.assertFalse(ret["OK"])
         self.file.LFN = "/a/b/c"
 
-        # # no ownerDN
-        # force no owner DN because it takes the one of the current user
-        self.request.OwnerDN = ""
```
Comment (chaen) [inline src/DIRAC/RequestManagementSystem/private/test/Test_RequestValidator.py:80]: Can you add a backward compatibility test, i.e. a request that has no `Owner` but an `OwnerDN`

File: src/DIRAC/TransformationSystem/Client/WorkflowTasks.py (hunk 10 above)
Comment (chaen) [reply src/DIRAC/TransformationSystem/Client/WorkflowTasks.py:109]: I believe we still need two ifferent "owner" there. It's actually a confusing naming, because `owner`  and `ownerDN` can be two different identities. It should rather be `owner` and `onBehalfOf` 

It was introduced here https://github.com/DIRACGrid/DIRAC/pull/3723

File: src/DIRAC/RequestManagementSystem/private/test/Test_RequestValidator.py (hunk 12 above)
Comment (fstagni) [reply src/DIRAC/RequestManagementSystem/private/test/Test_RequestValidator.py:80]: Is `RequestValidator` running on client side or on server side? or both? If it is running only on Client side then no need add a test here.

File: src/DIRAC/RequestManagementSystem/private/RequestValidator.py (hunk 11 above)
Comment (fstagni) [reply src/DIRAC/RequestManagementSystem/private/RequestValidator.py]: It adds if it is empty. 
I am not sure if it is even ever possible, but this is what is done in v8.0.

File: src/DIRAC/TransformationSystem/Client/WorkflowTasks.py (hunk 10 above)
Comment (fstagni) [reply src/DIRAC/TransformationSystem/Client/WorkflowTasks.py:109]: I am getting rather confused here. Does this mean that this "OwnerDN" (that you want to call "onBehalfOf") correspond to the "AuthorDN" found in TransformationDB?

Again, @andresailer might know better.

Comment (chaen) [reply src/DIRAC/TransformationSystem/Client/WorkflowTasks.py:109]: I think that's where it ends up indeed.
It's easy to try though

File: src/DIRAC/RequestManagementSystem/private/RequestValidator.py (hunk 11 above)
Comment (chaen) [reply src/DIRAC/RequestManagementSystem/private/RequestValidator.py]: Yes of course, this is to set the ownerDN from the credentials (if not set) on the server side when uploading the request.  So here, you want to set just the `Owner` I guess.

File: src/DIRAC/RequestManagementSystem/private/test/Test_RequestValidator.py (hunk 12 above)
Comment (chaen) [reply src/DIRAC/RequestManagementSystem/private/test/Test_RequestValidator.py:80]: It's running on both sides

File: src/DIRAC/RequestManagementSystem/private/RequestValidator.py (hunk 11 above)
Comment (fstagni) [reply src/DIRAC/RequestManagementSystem/private/RequestValidator.py]: Well I can just simply remove the 3 lines above then...?

File: src/DIRAC/RequestManagementSystem/private/test/Test_RequestValidator.py (hunk 12 above)
Comment (fstagni) [reply src/DIRAC/RequestManagementSystem/private/test/Test_RequestValidator.py:80]: The unit test is there, added. There is also a system test in https://github.com/DIRACGrid/DIRAC/blob/integration/tests/System/rms_scripts.sh

File: src/DIRAC/TransformationSystem/Client/WorkflowTasks.py (hunk 10 above)
Comment (fstagni) [reply src/DIRAC/TransformationSystem/Client/WorkflowTasks.py:109]: Maybe the solution is to start by storing in TransformationDB an "OnBehalfOf" field that would store the username, and not the DN of the user. But this won't be done in this PR. 
I will see what can be done, but anyway the TS on the DIRAC certification setup is not using the ShifterProxy so it should be possible to test this already.

File: src/DIRAC/RequestManagementSystem/private/RequestValidator.py (hunk 11 above)
Comment (fstagni) [reply src/DIRAC/RequestManagementSystem/private/RequestValidator.py]: After applying this PR as patch to the DIRAC certification setup, and having run some integration tests, I am somewhat confident that the changes in this module cover our cases.

**Comment by fstagni** [comment]
> > You mean, merge as-is, try it out, and go back to the comments if needed?
> 
> no no, not merge, just `git apply` it

The PR content was tested at the recent hackathon (mostly server side). It was verified that old requests can be properly treated. Not everything was tested, specifically the interactions with TS was not done (see https://github.com/DIRACGrid/DIRAC/pull/6653#discussion_r1134305673). It'd be a bit cumbersome to have more tests without having this merged first.

File: src/DIRAC/TransformationSystem/Client/WorkflowTasks.py (hunk 10 above)
Comment (andresailer) [reply src/DIRAC/TransformationSystem/Client/WorkflowTasks.py:109]: Definitely the TransformationDB should store Author(UserName) instead of AuthorDN.

The Owner, OwnerGroup, OwnerDN should be consistent though. Either all belong to the shifterProxy, or be the author of the transformation.

**Comment by fstagni** [comment]
> > > You mean, merge as-is, try it out, and go back to the comments if needed?
> > 
> > 
> > no no, not merge, just `git apply` it
> 
> The PR content was tested at the recent hackathon (mostly server side). It was verified that old requests can be properly treated. Not everything was tested, specifically the interactions with TS was not done (see [#6653 (comment)](https://github.com/DIRACGrid/DIRAC/pull/6653#discussion_r1134305673)). It'd be a bit cumbersome to have more tests without having this merged first.

Review, please. I would like to merge this one before the next hackathon.

File: src/DIRAC/TransformationSystem/Client/WorkflowTasks.py (hunk 10 above)
Comment (chaen) [reply src/DIRAC/TransformationSystem/Client/WorkflowTasks.py:109]: @andresailer could you come up with an integration test for that specific case ?

Comment (andresailer) [reply src/DIRAC/TransformationSystem/Client/WorkflowTasks.py:109]: Which specific case do you mean?

Comment (chaen) [reply src/DIRAC/TransformationSystem/Client/WorkflowTasks.py:109]: The case where this was designed for, that is for multiVO TS (you introduced it in #3723)

**Comment by fstagni** [comment]
PR fully tested in latest hackathon, no issues found.

File: src/DIRAC/TransformationSystem/Client/WorkflowTasks.py (hunk 10 above)
Comment (chaen) [reply src/DIRAC/TransformationSystem/Client/WorkflowTasks.py:109]: After discussing with @andresailer it seems the intention was to have the 3 (DN, Owner, Group) always consistent, but that was never checked, so not clear we actually need an extra test

**Comment by fstagni** [comment]
Before I forget how I wrote this PR, last chance to review it.

**Comment by fstagni** [comment]
As agreed during today's BiLD meeting, this PR will be merged on Tuesday 30th even in absence of a positive review.

**Comment by fstagni** [comment]
> As agreed during today's BiLD meeting, this PR will be merged on Tuesday 30th even in absence of a positive review.

Merging in 1 hour
